File: src/GLFMmatlab/GLFM_infer.m
function [hidden,data_transformed] = GLFM_infer(data,varargin)
    % Wrapper .m function to call .cpp MATLAB wrapper (simplifies call)
    % Three possible calls:
    %           hidden = GLFM_infer(data)
    %           hidden = GLFM_infer(data,hidden)
    %           hidden = GLFM_infer(data,hidden,params)
    %
    %   Inputs:
    %       data: structure with all input data information
    %           (*) data.X: N*D observation matrix (raw)
    %           (*) data.C: 1*D string array with input data types
    %       (*) mandatory
    %
    %       ------------- optional ---------------------
    %
    %       hidden: hidden structure to initialize inference algorithm
    %             hidden.Z: feature assignment N*K matrix
    %       params: structure with simulation parameters and hyperparameters
    %   Outputs:
    %       hidden: hidden structure with samples from posterior, as well as
    %               inner transformation parameters
    %           hidden.Z    : N*K feature activation matrix
    %           hidden.B    : K*D feature effect matrix
    %           hidden.theta: D*maxR matrix with aux. variables for ordinal data
    %           hidden.mu   : 1*D shift parameter for inner transformation
    %           hidden.w    : 1*D scale parameter for inner transformation
    %           hidden.s2y  : 1*D per-dim inferred noise variance for pseudo-obs

    switch length(varargin)
        case 0
            hidden = [];
            % initialize default values for params
            params = init_default_params(data, []);

        case 1, hidden = varargin{1};
            params = init_default_params(data, []);

        case 2, hidden = varargin{1}; params = varargin{2};
            params = init_default_params(data, params); % eventually complete params structure

        otherwise
            error('Incorrect number of input parameters: should be 0, 1 or 2');
    end

    D = size(data.X,2);
    N = size(data.X,1);

    % initialize Z if necessary
    if isempty(hidden)
        clearvars hidden
        hidden.Z = double(rand(N,2)>0.8);
        if (params.bias == 1)
            hidden.Z = [ones(N,1) hidden.Z];
        elseif (params.bias > 1)
            error('There is more than 1 bias specified, but structure Z has not been initialized...');
        end
    end

    % replace missings
    data_transformed = data;
    data_transformed.X(isnan(data_transformed.X)) = params.missing;

    % change labels for categorical and ordinal vars such that > 0
    %V_offset = zeros(1,D);
    for d=1:D
        if (data_transformed.C(d) == 'c') || (data_transformed.C(d) == 'o')
            mask = data_transformed.X(:,d) ~= params.missing;
            %V_offset(d) = min( data_transformed.X(mask,d) );
            %data_transformed.X(mask,d) = data_transformed.X(mask,d) - V_offset(d) + 1;
            uniqueVal= unique(data_transformed.X(mask,d));
            Xaux=[];
            for i=1:length(uniqueVal)
                Xaux(data_transformed.X(:,d)==uniqueVal(i)) = i;
            end
            Xaux(~mask)=params.missing;
            data_transformed.X(:,d)= Xaux;
        end
    end

    % eventually, apply external transform
    for r=1:size(data_transformed.X,2)
        if ~isempty(params.t{r})
            data_transformed.X(:,r) = params.t_1{r}(data_transformed.X(:,r)); % work in logarithm space better
            data_transformed.C(r) = params.ext_dataType{r};
        end
    end

    func = 1*ones(1,size(data_transformed.X,2)); % type of internal transformation for positive real-valued data

    %% call .cpp wrapper function
    tic;
    [Z B theta mu w s2Y]= IBPsampler(data_transformed.X,data_transformed.C, hidden.Z, params.bias, func, ...
        params.s2u, params.s2B, params.alpha, params.Niter, params.maxK, params.missing);
    hidden.time = toc;
    fprintf('Elapsed time %.2f seconds.', hidden.time );

    %% prepare output structure
    hidden.Z = Z'; % it returns a K*N matrix, should be inverted
    hidden.B = B;
    hidden.theta = theta;
    hidden.mu = mu; % mapping functions info (necessary for plotting)
    hidden.w = w;
    hidden.s2Y = s2Y;
    hidden.R = ones(1,D);
    for d=1:D
        if (data_transformed.C(d) == 'c') || (data_transformed.C(d) == 'o')
            hidden.R(d) = length(unique(data_transformed.X(data_transformed.X(:,d)~= params.missing,d) ));
        end
    end
    if params.verbose
        hidden.time
        sum(hidden.Z)
    end
end

File: src/GLFMR/GLFM_infer.R
#' @description Wrapper .R function to call .cpp R wrapper
#' Inputs:
#'@param data is a list with X and C
#'@param X: N*D data matrix X
#'@param C: 1xD string with data types, D = number of dimensions
# ----------------(optional) ------------------
#' @param varargin is a list of lists containing 2 lists hidden and params 
#'@param hidden: list with the following latent variables (Z,B):
#'@param Z: NxK matrix of feature patterns
#'@param B: latent feature list with D matrices of size  K * maxR  where
#'@param D: number of dimensions
#'@param K: number of latent variables
#'@param maxR: maximum number of categories across all dimensions
#'@param params: list with parameters (mu,w,theta)
#'@param  mu: 1*D shift parameter
#'@param w: 1*D scale parameter
#'@param theta: D*maxR matrix of auxiliary vars (for ordinal variables)
#'@param s2y: 1*D per dim inferred noise variance for pseudo observations
#' @return A list of 3 lists (data, hidden, params) with inferred values

GLFM_infer<-function(data,varargin){
  require(matrixStats)
  source("aux/init_default_params.R")
  varargin_size<-length(varargin)
  if(varargin_size==0){
    Z<-c()
    params2 <- init_default_params(data, c())
  }else if(varargin_size==1){
      Z<-unlist(varargin[[1]])
      params2 <- init_default_params(data, c())
  }else if(varargin_size==2){
    Z<-unlist(varargin[[1]])
    params2 <- init_default_params(data, varargin[[2]])
  }else{
    stop("Incorrect number of input parameters: should be 0, 1 or 2")
  }
  D <- dim(data$X)[2]
  N <- dim(data$X)[1]
 
  if(length(Z)==0){
    m0<-matrix(0,N,2)
    Z <- apply(m0, c(1,2), function(x) sample(c(0,1),1,prob=c(0.8,0.2)))
  }
  if(params2$bias == 1 && length(params2$bias)>0){
    Z <-cbind(rep(1,N),Z)
  }
  else if(params2$bias>1  && length(params2$bias)>0){
    stop("There is more than 1 bias specified, but the structure Z has not been initialized") 
  }
  # replace missing values
  idx_missing<-which(is.nan(data$X))
  data$X[idx_missing] <- params2$missing
  idx_missing<-which((data$X)==params2$missing)
  X_aux<-data$X
  aa<-max(X_aux)
  X_aux[idx_missing] <- aa+1
  V_offset<-colMins(X_aux)
  V_offset_mat<-matrix(V_offset,nrow=N,ncol=D,byrow=TRUE)
  X_aux<-data$X-V_offset_mat+1
  idx_catord<-which(data$C=='c' | data$C=='o')
  if(length(idx_catord)>0){
    data$X[,idx_catord] <-X_aux[,idx_catord]
    bu<-apply(X_aux[,idx_catord,drop=FALSE], 2, function(x)length(unique(x)))
      idx_dat<-which(colMaxs(X_aux[,idx_catord,drop=FALSE])!=bu)
      if(length(idx_dat)>0){
      for(ii in 1:length(idx_dat)){
      idxs_bad<-which(X_aux[,idx_dat[ii]]>bu[idx_dat[ii]])
      while(length(idxs_bad)>0){
        X_aux[idxs_bad,idx_dat[ii]]<-X_aux[idxs_bad,idx_dat[ii]]-1
        idxs_bad<-which(X_aux[,idx_dat[ii]]>bu[idx_dat[ii]])
          }
        }
      }
    data$X[,idx_catord]<-X_aux[,idx_catord]
    data$X[idx_missing]<-params2$missing 
  }
  
  if( "transf_dummie" %in% names(params2)){
    if(params2$transf_dummie){
      if(is.list(params2$t_1)==FALSE){
     data$X[,params2$idx_transform]<-params2$t_1(data$X[,params2$idx_transform])
     data$C[params2$idx_transform] <-params2$ext_datatype
      }else{
        for(ell in 1:length(params2$t_1)){
          data$X[,params2$idx_transform[[ell]]]<-params2$t_1[[ell]](data$X[,params2$idx_transform[[ell]]])
          data$C[params2$idx_transform[[ell]]] <-params2$ext_datatype[[ell]]
          }
      }
    }
  }
  
  func_bit<-rep(1,dim(data$X)[2])
  setwd("../Ccode/")
  library(RcppGSLExample)
  # call .Rcpp wrapper function
  hidden<-IBPsampler(t(data$X),(data$C),t(Z),params2$bias,func_bit,params2$s2u,params2$s2B,params2$alpha,params2$Niter,params2$maxK,params2$missing)
  R<-rep(1,D)
  if(length(idx_catord)>0){
    X_aux<-data$X
    aa<-min(X_aux)
    X_aux[idx_missing] <- aa-1
    V_offset<--colMins(-X_aux)
    R[idx_catord]<-V_offset[idx_catord]
  }
  hidden$Z<-t(hidden$Z)
  hidden<-append(hidden, list("R"=R))
  setwd("../GLFMR")
  return(list("data"=data,"hidden"=hidden,"params"=params2))
}
 
  
  
  
  